APIv4 - Use empty string instead of 'null' to pass null values to the db

colemanw opened the pull request:
Overview
----------------------------------------
A less hacky way to treat NULL values in APIv4

Before
----------------------------------------
Passing the string `'null'` to represent null values.

After
----------------------------------------
Passing `''` to represent null values.

Technical Details
----------------------------------------
`CRM_Core_DAO::copyValues()` already converts `''` to `'null'`, and passing the string `'null'` into `BAO::create` functions prematurely can introduce subtle bugs because the string `'null'` is `!empty()` whereas `''` is `empty()`.

colemanw committed:
APIv4 - Use empty string instead of 'null' to pass null values to the database

`CRM_Core_DAO::copyValues()` already converts '' to 'null', and passing the string 'null'
into `BAO::writeRecords` or `BAO::create` functions prematurely can introduce subtle bugs
because the string 'null' is `!empty()` whereas '' is `empty()`.

eileenmcnaughton commented:
@colemanw I think this relates

<p style="box-sizing: inherit; font-family: Roboto, sans-serif; font-size: var(--font-size-sm); line-height: var(--line-height-base); color: rgb(68, 68, 68); outline: 0px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span style="box-sizing: inherit; outline: 0px; font-family: Roboto, sans-serif !important; font-weight: bold;">api_v3_ContactTest.testContactGreetingsCreateWithCustomField with data set "APIv4"</span><span> </span>(from api_v3_ContactTest__testContactGreetingsCreateWithCustomField)</p><div style="box-sizing: inherit; outline: 0px; font-family: Roboto, sans-serif; color: rgb(68, 68, 68); font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; text-align: right;">Failing for the past 1 build (Since<span> </span><a href="https://test.civicrm.org/job/CiviCRM-Core-PR/46639/" class="model-link" style="box-sizing: inherit; text-decoration: none !important; font-weight: var(--link-font-weight); color: rgb(38, 166, 154); outline: 0px; font-family: Roboto, sans-serif !important;"><span class="build-status-icon__wrapper icon-red icon-sm" style="box-sizing: inherit; width: 16px; height: 16px; color: var(--danger); fill: var(--danger); opacity: 1; display: inline-block; position: relative; outline: 0px; font-family: Roboto, sans-serif !important;"><span class="build-status-icon__outer" style="box-sizing: inherit; display: block; position: absolute; top: 8px; left: 8px; width: 16px; height: 16px; transform: translate(-50%, -50%); outline: 0px; font-family: Roboto, sans-serif !important;"><svg viewBox="0 0 24 24" focusable="false" class="svg-icon "><use href="/images/build-status/build-status-sprite.svg#build-status-static"></use></svg></span><svg viewBox="0 0 24 24" focusable="false" class="svg-icon icon-red icon-sm"><use href="/static/97ea3670/images/build-status/build-status-sprite.svg#last-failed"></use></svg></span>#46639</a><span> </span>)</div><div style="box-sizing: inherit; outline: 0px; font-family: Roboto, sans-serif; color: rgb(68, 68, 68); font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; text-align: right;"><a href="https://test.civicrm.org/job/CiviCRM-Core-PR/46639/testReport/junit/(root)/api_v3_ContactTest/testContactGreetingsCreateWithCustomField_with_data_set__APIv4_/history" style="box-sizing: inherit; text-decoration: none !important; font-weight: var(--link-font-weight); color: rgb(38, 166, 154); outline: 0px; font-family: Roboto, sans-serif !important;">Took 0.9 sec.</a></div><div id="description" style="box-sizing: inherit; outline: 0px; font-family: Roboto, sans-serif; color: rgb(68, 68, 68); font-size: 13px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div style="box-sizing: inherit; outline: 0px; font-family: Roboto, sans-serif !important;"></div></div>
<h3 style="box-sizing: inherit; line-height: var(--line-height-heading); font-weight: bold; display: block; margin-top: 0.5rem; margin-bottom: 0.5rem; font-size: 1.5rem; outline: 0px; font-family: Roboto, sans-serif; color: rgb(68, 68, 68); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Stacktrace</h3><pre style="box-sizing: inherit; font-size: var(--font-size-monospace); white-space: pre-wrap; overflow-wrap: break-word; margin: 0px; outline: 0px; font-family: &quot;Roboto Mono&quot;, monospace !important; color: rgb(68, 68, 68); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">api_v3_ContactTest::testContactGreetingsCreateWithCustomField with data set "APIv4" (4)
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-''
+'Dear Alan Mice'

/home/jenkins/bknix-dfl/build/core-22587-5t46z/web/sites/all/modules/civicrm/tests/phpunit/api/v3/ContactTest.php:4869
/home/jenkins/bknix-dfl/build/core-22587-5t46z/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:262
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671</pre>

colemanw committed:
Contact Greetings - Fix assumption about empty values

The function `ensureGreetingParamsAreSet` was assuming that the string 'null' means NULL
but an empty string was "no value given". In reality, an empty string can also signify NULL,
and only a true NULL or unset value means "no value given".

colemanw commented:
@eileenmcnaughton got it passing!
I think this PR is going to save us a lot of headaches in the long run. There are hundreds of potential `if (empty('null'))` gotchas in our BAOs that this will avoid.

monishdeb commented:
Looks good to me. r-run passed
